File: src/Application Importer/Validator.cs
using System;
using System.Collections;
using System.Data;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace FSM.Importer
{
    public class Validator
    {

        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        private ArrayList _DataToValidate;
        private ArrayList _arrayList;
        private FRMPartsImport _fRMImportNew;

        public Validator(ArrayList arrayList, FRMPartsImport fRMImportNew)
        {
            // TODO: Complete member initialization
            _arrayList = arrayList;
            _fRMImportNew = fRMImportNew;
        }

        public ArrayList DataToValidate
        {
            get
            {
                return _DataToValidate;
            }

            set
            {
                _DataToValidate = value;
            }
        }

        private ArrayList _Errors;

        public ArrayList Errors
        {
            get
            {
                return _Errors;
            }

            set
            {
                _Errors = value;
            }
        }

        private FRMImport _ImportForm;

        public FRMImport ImportForm
        {
            get
            {
                return _ImportForm;
            }

            set
            {
                _ImportForm = value;
            }
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        public Validator(ArrayList DataToValidateIn, ref FRMImport ImportFormIn)
        {
            DataToValidate = DataToValidateIn;
            Errors = new ArrayList();
            ImportForm = ImportFormIn;
            ImportForm.MoveProgressOn();
        }

        public ArrayList Validate(int SupplierID)
        {
            ArrayList returnArr = null;
            ResetCells();
            for (int dvIndex = 0, loopTo = DataToValidate.Count - 1; dvIndex <= loopTo; dvIndex++)
            {
                DataView dv = (DataView)DataToValidate[dvIndex];
                if (dvIndex == 0 & (dv.Table.TableName ?? "") == "Parts")
                {

                    // Check Columns
                    var ExpectedColumns = new string[] { "Part Code", "Name/Description", "Category", "Notes", "Sell Price", "Supplier Part Code", "Supplier Qty", "Supplier Price" };
                    for (int ExpectedColumnIndex = 0, loopTo1 = ExpectedColumns.Length - 1; ExpectedColumnIndex <= loopTo1; ExpectedColumnIndex++)
                    {
                        string ExpectedColumn = ExpectedColumns[ExpectedColumnIndex];
                        if (!((dv.Table.Columns[ExpectedColumnIndex].ColumnName ?? "") == (ExpectedColumn ?? "")))
                        {
                            Errors.Add(string.Format("{0} Column is Missing at position {1}", ExpectedColumn, ExpectedColumnIndex));
                        }
                    }

                    if (Errors.Count == 0)
                    {
                        // Check Data
                        returnArr = ValidateParts(dv, SupplierID);
                    }
                }
                else if (dvIndex == 0 & (dv.Table.TableName ?? "") == "Sites")
                {

                    // Check Data
                    returnArr = ValidateSites(dv);
                }
                else
                {
                    Errors.Add("The first worksheet must be titled \"Parts\"" + "Or \"Sites\"");
                }

                ImportForm.MoveProgressOn();
            }

            return returnArr;
        }

        public void ResetCells()
        {
            // For Each tp As TabPage In ImportForm.tcData.TabPages
            // For Each row As DataRow In CType(tp.Controls(0), UCData).Data.Table.Rows
            // For Each col As DataColumn In CType(tp.Controls(0), UCData).Data.Table.Columns
            // If col.ColumnName.EndsWith("COLOURBOOLEANCOLUMN") Then
            // row.Item(col.ColumnName) = False
            // End If
            // Next
            // Next
            // Next
        }

        public void ErrorInCell(int SheetNumber, int RowNumber, string ColumnName, string Message)
        {
            // CType(ImportForm.tcData.TabPages(SheetNumber).Controls(0), UCData).Data.Table.Rows(RowNumber).Item(ColumnName & "COLOURBOOLEANCOLUMN") = True

            Errors.Add(string.Format("Error in {0} at row {1}: {2} - {3}", ((DataView)DataToValidate[SheetNumber]).Table.TableName, RowNumber + 1, ColumnName, Message));
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        private ArrayList ValidateParts(DataView dvParts, int SupplierID)
        {
            int insertCount = 0;
            int updateCount = 0;
            try
            {
                for (int partIndex = 0, loopTo = dvParts.Count - 1; partIndex <= loopTo; partIndex++)
                {
                    var part = dvParts[partIndex];

                    // ImportForm.SetLastPartAttempted(part("Part Code"))

                    // Part Code
                    if (Information.IsDBNull(part["Part Code"]) || Conversions.ToString(part["Part Code"]).Length <= 0)
                    {
                        ErrorInCell(0, partIndex, "Part Code (MPN)", "Required field");
                    }
                    else
                    {
                        int partsFound = App.DB.Part.Part_Exists(Conversions.ToString(part["Part Code"]), Conversions.ToString(part["Supplier Part Code"]));
                        if (partsFound == 0)
                        {
                            insertCount += 1;

                            // Name/Description
                            if (Information.IsDBNull(part["Name/Description"]) || Conversions.ToString(part["Name/Description"]).Length <= 0)
                            {
                                ErrorInCell(0, partIndex, "Name/Description", "Required field for an insert");
                            }

                            // Category
                            if (Information.IsDBNull(part["Category"]) || Conversions.ToString(part["Category"]).Length <= 0)
                            {
                                ErrorInCell(0, partIndex, "Category", "Required field for an insert");
                            }
                            else if (App.DB.ImportValidation.CategoryExists(Conversions.ToString(part["Category"])) == false)
                            {
                                ErrorInCell(0, partIndex, "Category", Conversions.ToString("'" + part["Category"] + "' does not exist"));
                            }

                            if (!Information.IsDBNull(part["Sell Price"]))
                            {
                                if (!(part["Sell Price"].GetType() == typeof(decimal)))
                                {
                                    ErrorInCell(0, partIndex, "Sell Price", "Required Decimal field for an insert - e.g 1.23");
                                }
                            }

                            // part code (S1)
                            if (Information.IsDBNull(part["Supplier Part Code"]) || Conversions.ToString(part["Supplier Part Code"]).Length <= 0)
                            {
                                ErrorInCell(0, partIndex, "Supplier Part Code (SPN)", "Required for an insert");
                            }

                            // qty (S1)
                            if (Information.IsDBNull(part["Supplier Qty"]) || !(part["Supplier Qty"].GetType() == typeof(double)))
                            {
                                ErrorInCell(0, partIndex, "Supplier Qty", "Required Double field for an insert - e.g 1.23");
                            }

                            // Price (S1)
                            if (Information.IsDBNull(part["Supplier Price"]) || !(part["Supplier Price"].GetType() == typeof(decimal)))
                            {
                                ErrorInCell(0, partIndex, "Supplier Price", "Reqired Decimal field for an insert - e.g £1.23");
                            }
                        }
                        else
                        {
                            updateCount += 1;
                            int partID = partsFound;

                            // Category
                            if (!Information.IsDBNull(part["Category"]) && Conversions.ToString(part["Category"]).Length > 0)
                            {
                                if (App.DB.ImportValidation.CategoryExists(Conversions.ToString(part["Category"])) == false)
                                {
                                    ErrorInCell(0, partIndex, "Category", Conversions.ToString("'" + part["Category"] + "' does not exist"));
                                }
                            }

                            if (!Information.IsDBNull(part["Sell Price"]))
                            {
                                if (!(part["Sell Price"].GetType() == typeof(decimal)))
                                {
                                    ErrorInCell(0, partIndex, "Sell Price", "Required Decimal field for an update - e.g 1.23");
                                }
                            }

                            Array suppliersFound = App.DB.PartSupplier.Get_ByPartID(partID).Table.Select("SupplierID = " + SupplierID);
                            if (suppliersFound.Length == 0)
                            {
                                // part code (S1)
                                if (Information.IsDBNull(part["Supplier Part Code"]) || Conversions.ToString(part["Supplier Part Code"]).Length <= 0)
                                {
                                    ErrorInCell(0, partIndex, "Supplier Part Code (SPN)", "Required field for an update when new part supplier");
                                }

                                // qty (S1)
                                if (Information.IsDBNull(part["Supplier Qty"]) || !(part["Supplier Qty"].GetType() == typeof(double)))
                                {
                                    ErrorInCell(0, partIndex, "Supplier Qty", "Required Double field for an update when new part supplier - e.g 1.23");
                                }

                                // Price (S1)
                                if (Information.IsDBNull(part["Supplier Price"]) || !(part["Supplier Price"].GetType() == typeof(decimal)))
                                {
                                    ErrorInCell(0, partIndex, "Supplier Price", "Reqired Decimal field for an update when new part supplier - e.g £1.23");
                                }
                            }
                            else
                            {
                                // qty (S1)
                                if (!Information.IsDBNull(part["Supplier Qty"]))
                                {
                                    if (!(part["Supplier Qty"].GetType() == typeof(double)))
                                    {
                                        ErrorInCell(0, partIndex, "Supplier Qty", "Required Double field for an update when updating part supplier - e.g 1.23");
                                    }
                                }

                                // Price (S1)
                                if (!Information.IsDBNull(part["Supplier Price"]))
                                {
                                    if (!(part["Supplier Price"].GetType() == typeof(decimal)))
                                    {
                                        ErrorInCell(0, partIndex, "Supplier Price", "Reqired Decimal field for an update when updating part supplier - e.g £1.23");
                                    }
                                }
                            }
                        }
                    }

                    ImportForm.MoveProgressOn();
                }

                var returnArr = new ArrayList();
                returnArr.Add(insertCount);
                returnArr.Add(updateCount);
                return returnArr;
            }
            catch (Exception ex)
            {
                Errors.Add("Error occured while validating" + Constants.vbCrLf + ex.Message);
                return null;
            }
        }

        private ArrayList ValidateSites(DataView dvSites)
        {
            int cnt = 0;
            try
            {
                for (int siteIndex = 0, loopTo = dvSites.Count - 1; siteIndex <= loopTo; siteIndex++)
                {
                    var site = dvSites[siteIndex];

                    // URPN
                    if (Information.IsDBNull(site["PolicyNumber"]) || Conversions.ToString(site["PolicyNumber"]).Length <= 0)
                    {
                        ErrorInCell(0, siteIndex, "URPN", "Required field");
                    }
                    else if (Entity.Sys.Helper.MakeStringValid(site["Address1"]).Trim().Length == 0)
                    {
                        ErrorInCell(0, siteIndex, "Address1", "Required field for an insert");
                        // If Entity.Sys.Helper.MakeStringValid(site("Postcode")).Trim.Length = 0 Then
                        // ErrorInCell(0, siteIndex, "Postcode", "Required field for an insert")
                        // End If

                    }

                    cnt += 1;
                    ImportForm.MoveProgressOn();
                }

                var returnArr = new ArrayList();
                returnArr.Add(cnt);
                return returnArr;
            }
            catch (Exception ex)
            {
                Errors.Add("Error occured while validating" + Constants.vbCrLf + ex.Message);
                return null;
            }
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
    }
}

File: src/Application Importer/Validator.vb
Imports System.Data.SqlClient

Namespace Importer

    Public Class Validator

#Region "Properties"

        Private _DataToValidate As ArrayList
        Private _arrayList As ArrayList
        Private _fRMImportNew As FRMPartsImport

        Sub New(arrayList As ArrayList, fRMImportNew As FRMPartsImport)
            ' TODO: Complete member initialization
            _arrayList = arrayList
            _fRMImportNew = fRMImportNew
        End Sub

        Public Property DataToValidate() As ArrayList
            Get
                Return _DataToValidate
            End Get
            Set(ByVal Value As ArrayList)
                _DataToValidate = Value
            End Set
        End Property

        Private _Errors As ArrayList

        Public Property Errors() As ArrayList
            Get
                Return _Errors
            End Get
            Set(ByVal Value As ArrayList)
                _Errors = Value
            End Set
        End Property

        Private _ImportForm As FRMImport

        Public Property ImportForm() As FRMImport
            Get
                Return _ImportForm
            End Get
            Set(ByVal Value As FRMImport)
                _ImportForm = Value
            End Set
        End Property

#End Region

#Region "Generic Functions"

        Public Sub New(ByVal DataToValidateIn As ArrayList, ByRef ImportFormIn As FRMImport)
            DataToValidate = DataToValidateIn
            Errors = New ArrayList
            ImportForm = ImportFormIn

            ImportForm.MoveProgressOn()
        End Sub

        Public Function Validate(ByVal SupplierID As Integer) As ArrayList
            Dim returnArr As ArrayList = Nothing

            ResetCells()

            For dvIndex As Integer = 0 To DataToValidate.Count - 1
                Dim dv As DataView = CType(DataToValidate(dvIndex), DataView)

                If dvIndex = 0 And dv.Table.TableName = "Parts" Then

                    'Check Columns
                    Dim ExpectedColumns() As String = New String() {"Part Code", "Name/Description", "Category", "Notes", "Sell Price", "Supplier Part Code", "Supplier Qty", "Supplier Price"}
                    For ExpectedColumnIndex As Integer = 0 To ExpectedColumns.Length - 1
                        Dim ExpectedColumn As String = ExpectedColumns(ExpectedColumnIndex)

                        If Not dv.Table.Columns(ExpectedColumnIndex).ColumnName = ExpectedColumn Then
                            Errors.Add(String.Format("{0} Column is Missing at position {1}", ExpectedColumn, ExpectedColumnIndex))
                        End If
                    Next

                    If Errors.Count = 0 Then
                        'Check Data
                        returnArr = ValidateParts(dv, SupplierID)
                    End If
                ElseIf dvIndex = 0 And dv.Table.TableName = "Sites" Then

                    'Check Data
                    returnArr = ValidateSites(dv)
                Else
                    Errors.Add("The first worksheet must be titled ""Parts""" & "Or ""Sites""")
                End If

                ImportForm.MoveProgressOn()
            Next

            Return returnArr
        End Function

        Public Sub ResetCells()
            'For Each tp As TabPage In ImportForm.tcData.TabPages
            '    For Each row As DataRow In CType(tp.Controls(0), UCData).Data.Table.Rows
            '        For Each col As DataColumn In CType(tp.Controls(0), UCData).Data.Table.Columns
            '            If col.ColumnName.EndsWith("COLOURBOOLEANCOLUMN") Then
            '                row.Item(col.ColumnName) = False
            '            End If
            '        Next
            '    Next
            'Next
        End Sub

        Public Sub ErrorInCell(ByVal SheetNumber As Integer, ByVal RowNumber As Integer, ByVal ColumnName As String, ByVal Message As String)
            'CType(ImportForm.tcData.TabPages(SheetNumber).Controls(0), UCData).Data.Table.Rows(RowNumber).Item(ColumnName & "COLOURBOOLEANCOLUMN") = True

            Errors.Add(String.Format("Error in {0} at row {1}: {2} - {3}", CType(DataToValidate(SheetNumber), DataView).Table.TableName, RowNumber + 1, ColumnName, Message))
        End Sub

#End Region

#Region "Private Functions"

        Private Function ValidateParts(ByVal dvParts As DataView, ByVal SupplierID As Integer) As ArrayList
            Dim insertCount As Integer = 0
            Dim updateCount As Integer = 0

            Try
                For partIndex As Integer = 0 To dvParts.Count - 1
                    Dim part As DataRowView = dvParts(partIndex)

                    'ImportForm.SetLastPartAttempted(part("Part Code"))

                    'Part Code
                    If IsDBNull(part("Part Code")) OrElse CStr(part("Part Code")).Length <= 0 Then
                        ErrorInCell(0, partIndex, "Part Code (MPN)", "Required field")
                    Else

                        Dim partsFound As Integer = DB.Part.Part_Exists(part("Part Code"), part("Supplier Part Code"))

                        If partsFound = 0 Then
                            insertCount += 1

                            'Name/Description
                            If IsDBNull(part("Name/Description")) OrElse CStr(part("Name/Description")).Length <= 0 Then
                                ErrorInCell(0, partIndex, "Name/Description", "Required field for an insert")
                            End If

                            'Category
                            If IsDBNull(part("Category")) OrElse CStr(part("Category")).Length <= 0 Then
                                ErrorInCell(0, partIndex, "Category", "Required field for an insert")
                            ElseIf DB.ImportValidation.CategoryExists(part("Category")) = False Then
                                ErrorInCell(0, partIndex, "Category", "'" & part("Category") & "' does not exist")
                            End If

                            If Not IsDBNull(part("Sell Price")) Then
                                If Not part("Sell Price").GetType() Is GetType(Decimal) Then
                                    ErrorInCell(0, partIndex, "Sell Price", "Required Decimal field for an insert - e.g 1.23")
                                End If
                            End If

                            'part code (S1)
                            If IsDBNull(part("Supplier Part Code")) OrElse CStr(part("Supplier Part Code")).Length <= 0 Then
                                ErrorInCell(0, partIndex, "Supplier Part Code (SPN)", "Required for an insert")
                            End If

                            'qty (S1)
                            If IsDBNull(part("Supplier Qty")) OrElse Not part("Supplier Qty").GetType() Is GetType(Double) Then
                                ErrorInCell(0, partIndex, "Supplier Qty", "Required Double field for an insert - e.g 1.23")
                            End If

                            'Price (S1)
                            If IsDBNull(part("Supplier Price")) OrElse Not part("Supplier Price").GetType() Is GetType(Decimal) Then
                                ErrorInCell(0, partIndex, "Supplier Price", "Reqired Decimal field for an insert - e.g £1.23")
                            End If
                        Else
                            updateCount += 1

                            Dim partID As Integer = partsFound

                            'Category
                            If Not IsDBNull(part("Category")) AndAlso CStr(part("Category")).Length > 0 Then
                                If DB.ImportValidation.CategoryExists(part("Category")) = False Then
                                    ErrorInCell(0, partIndex, "Category", "'" & part("Category") & "' does not exist")
                                End If
                            End If

                            If Not IsDBNull(part("Sell Price")) Then
                                If Not part("Sell Price").GetType() Is GetType(Decimal) Then
                                    ErrorInCell(0, partIndex, "Sell Price", "Required Decimal field for an update - e.g 1.23")
                                End If
                            End If

                            Dim suppliersFound As Array = DB.PartSupplier.Get_ByPartID(partID).Table.Select("SupplierID = " & SupplierID)

                            If suppliersFound.Length = 0 Then
                                'part code (S1)
                                If IsDBNull(part("Supplier Part Code")) OrElse CStr(part("Supplier Part Code")).Length <= 0 Then
                                    ErrorInCell(0, partIndex, "Supplier Part Code (SPN)", "Required field for an update when new part supplier")
                                End If

                                'qty (S1)
                                If IsDBNull(part("Supplier Qty")) OrElse Not part("Supplier Qty").GetType() Is GetType(Double) Then
                                    ErrorInCell(0, partIndex, "Supplier Qty", "Required Double field for an update when new part supplier - e.g 1.23")
                                End If

                                'Price (S1)
                                If IsDBNull(part("Supplier Price")) OrElse Not part("Supplier Price").GetType() Is GetType(Decimal) Then
                                    ErrorInCell(0, partIndex, "Supplier Price", "Reqired Decimal field for an update when new part supplier - e.g £1.23")
                                End If
                            Else
                                'qty (S1)
                                If Not IsDBNull(part("Supplier Qty")) Then
                                    If Not part("Supplier Qty").GetType() Is GetType(Double) Then
                                        ErrorInCell(0, partIndex, "Supplier Qty", "Required Double field for an update when updating part supplier - e.g 1.23")
                                    End If
                                End If

                                'Price (S1)
                                If Not IsDBNull(part("Supplier Price")) Then
                                    If Not part("Supplier Price").GetType() Is GetType(Decimal) Then
                                        ErrorInCell(0, partIndex, "Supplier Price", "Reqired Decimal field for an update when updating part supplier - e.g £1.23")
                                    End If
                                End If
                            End If
                        End If
                    End If

                    ImportForm.MoveProgressOn()
                Next

                Dim returnArr As New ArrayList
                returnArr.Add(insertCount)
                returnArr.Add(updateCount)
                Return returnArr
            Catch ex As Exception
                Errors.Add("Error occured while validating" & vbCrLf & ex.Message)

                Return Nothing
            End Try
        End Function

        Private Function ValidateSites(ByVal dvSites As DataView) As ArrayList
            Dim cnt As Integer = 0

            Try

                For siteIndex As Integer = 0 To dvSites.Count - 1
                    Dim site As DataRowView = dvSites(siteIndex)

                    'URPN
                    If IsDBNull(site("PolicyNumber")) OrElse CStr(site("PolicyNumber")).Length <= 0 Then
                        ErrorInCell(0, siteIndex, "URPN", "Required field")
                    Else

                        If Entity.Sys.Helper.MakeStringValid(site("Address1")).Trim.Length = 0 Then
                            ErrorInCell(0, siteIndex, "Address1", "Required field for an insert")
                        End If
                        'If Entity.Sys.Helper.MakeStringValid(site("Postcode")).Trim.Length = 0 Then
                        '    ErrorInCell(0, siteIndex, "Postcode", "Required field for an insert")
                        'End If

                    End If
                    cnt += 1
                    ImportForm.MoveProgressOn()
                Next

                Dim returnArr As New ArrayList
                returnArr.Add(cnt)
                Return returnArr
            Catch ex As Exception
                Errors.Add("Error occured while validating" & vbCrLf & ex.Message)

                Return Nothing
            End Try
        End Function

#End Region

    End Class

End Namespace